docs/source/clickhouse/schema_design.rst

```diff
@@ -16,31 +18,37 @@ data in a column that contains two ``Nested`` arrays where the first array conta
 of the dictionary and the second array contains the values. This works well when
 your dataset design gives you the ability to filter based on other attributes to a small
 number of rows and the flexibility of completely arbitrary keys and values are a real requirement.
-Often, however, a ClickHouse user is interested in rows that contain a specific tag key or a
-specific ``key=value`` pair. In this case one is often best-served by creating a top-level
-column and "promoting" the data to it [#dupe]_.
-
-This promotion is preferable on a couple dimensions:
-
-1. Even though it looks like a dictionary, our tags and contexts "maps" are actually arrays of keys and values.
-   With array accesses Clickhouse needs to locate the index of the key whose value needs to be looked up.
-   That is an O(N) lookup. The same key may be located at different indexes for each row.
-   Once the index is known, looking up its value is O(1) since you perform direct index access.
-   If the query has to lookup M rows, this becomes an O(M * N) in terms of time complexity.
-
-   With direct column access, there is no index lookup for a key. The overall complexity to lookup M rows
-   for "promoted columns" is O(M).
-2. Queries that filter based on an array column may have to load the entire column into memory
-   in order to do filtering. Adding a new index on top of a top-level column is a more
-   straightforward operation that guarantees more consistent performance outcomes vs. iterating
-   over an array
+Often, however, a ClickHouse user is predominantly interested in rows that contain a specific tag key or a
+specific ``key=value`` pair that occurs in a very large percentage of the table data (for example,
+``http_status_code=200`` or ``hasKey(http_status_code)``). This breaks any efficiency
+of the bloom-filter style index on dictionary columns (see :ref:`bloom`).
+In this case one is often best-served by creating a top-level column and "promoting"
+the data to it [#dupe]_.
+
+It is very efficient for queries to filter by a specific matching value that exists in
+a high number of rows for a given tag key being non-empty or key:value pair on a column
```

Comment:
Are you talking about using the bloom filter or the promoted column ? 
The problem is that neither is efficient. If the value has low selectivity (present in few rows) the search can be made efficient.

- low selectivity, key=value condition -> do not promote the column, use the bloom filter on the nested column
- low selectivity, key LIKE %something% -> bloom filter does not help, if it is a common condition, promoted column is a bit better. It gets better as the nested column increases in size
- high selectivity -> both are bad. you need to scan all the blocks. The promoted column reduces the amount of data to scan making things a bit better